pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -137,28 +141,72 @@ func (m *mergeQuerier) LabelNames() ([]string, storage.Warnings, error) {
 	return labelNames, warnings, nil
 }
 
-type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
+type stringSliceFunc func(context.Context, storage.Querier) ([]string, storage.Warnings, error)
+
+type stringSliceFuncJob struct {
+	querier  storage.Querier
+	tenantID string
+	result   []string
+	warnings storage.Warnings
+}
 
 // mergeDistinctStringSlice is aggregating results from stringSliceFunc calls
-// on a querier. It removes duplicates and sorts the result. It doesn't require
-// the output of the stringSliceFunc to be sorted, as results of LabelValues
-// are not sorted.
-//
-// TODO: Consider running stringSliceFunc calls concurrently
+// on per querier in parallel. It removes duplicates and sorts the result. It
+// doesn't require the output of the stringSliceFunc to be sorted, as results
+// of LabelValues are not sorted.
 func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+	var jobs = make([]interface{}, len(m.tenantIDs))
+
+	for pos := range m.tenantIDs {
+		jobs[pos] = &stringSliceFuncJob{
+			querier:  m.queriers[pos],
+			tenantID: m.tenantIDs[pos],
+		}
+	}
+
+	run := func(ctx context.Context, jobIntf interface{}) error {
+		// end early when context has ended
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		job, ok := jobIntf.(*stringSliceFuncJob)
+		if !ok {
+			return fmt.Errorf("unexpected type %T", jobIntf)
+		}
+
+		result, resultWarnings, err := f(ctx, job.querier)
+		if err != nil {
+			return errors.Wrapf(err, `error querying {%s="%s"}`, defaultTenantLabel, job.tenantID)
```

Comment:
I think old error message ("error querying tenant id...") was more clear. This looks like we're running this specific query, which is not true. Same comment applies at other places.

Reply:
I agree the old format in this place is clearer. I mainly used that format because I wanted to ensure we have a consistent format and it was easy to convert `label.Labels` to string.

Implement the old format for both in 833a7f9